File: widgetssdk/src/test/java/com/glia/widgets/filepreview/domain/PutImageFileToDownloadsUseCaseTest.java
package com.glia.widgets.filepreview.domain;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import android.graphics.Bitmap;

import com.glia.widgets.filepreview.data.GliaFileRepository;
import com.glia.widgets.filepreview.domain.exception.FileNameMissingException;
import com.glia.widgets.filepreview.domain.usecase.PutImageFileToDownloadsUseCase;

import org.junit.Before;
import org.junit.Test;

import io.reactivex.Completable;

public class PutImageFileToDownloadsUseCaseTest {

    private GliaFileRepository fileRepository;
    private PutImageFileToDownloadsUseCase useCase;

    @Before
    public void setUp() {
        fileRepository = mock(GliaFileRepository.class);
        useCase = new PutImageFileToDownloadsUseCase(fileRepository);
    }

    @Test
    public void execute_returnOnError_whenFileNameIsNull() {
        useCase.execute(null, BITMAP)
                .test()
                .assertError(FileNameMissingException.class);
    }

    @Test
    public void execute_returnOnError_whenFileNameIsEmpty() {
        useCase.execute(NAME_EMPTY, BITMAP)
                .test()
                .assertError(FileNameMissingException.class);
    }

    @Test
    public void execute_returnOnComplete() {
        when(fileRepository.putImageToDownloads(any(), any()))
                .thenReturn(Completable.complete());
        useCase.execute(NAME, BITMAP)
                .test()
                .assertComplete();
    }

    @Test
    public void execute_returnOnComplete_whenBitmapIsNull() {
        when(fileRepository.putImageToDownloads(any(), any()))
                .thenReturn(Completable.complete());
        useCase.execute(NAME, null)
                .test()
                .assertComplete();
    }

    private static final String NAME = "NAME";
    private static final String NAME_EMPTY = "";
    private static final Bitmap BITMAP = mock(Bitmap.class);
}

File: widgetssdk/src/test/java/com/glia/widgets/filepreview/domain/PutImageFileToDownloadsUseCaseTest.kt
package com.glia.widgets.filepreview.domain

import android.graphics.Bitmap
import com.glia.widgets.filepreview.data.GliaFileRepository
import com.glia.widgets.filepreview.domain.exception.FileNameMissingException
import com.glia.widgets.filepreview.domain.usecase.PutImageFileToDownloadsUseCase
import io.reactivex.rxjava3.core.Completable
import org.junit.Before
import org.junit.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever

class PutImageFileToDownloadsUseCaseTest {
    private lateinit var fileRepository: GliaFileRepository
    private lateinit var useCase: PutImageFileToDownloadsUseCase

    @Before
    fun setUp() {
        fileRepository = mock()
        useCase = PutImageFileToDownloadsUseCase(fileRepository)
    }

    @Test
    fun execute_emitsFileNameMissingException_whenFileNameIsNull() {
        useCase(null, BITMAP)
            .test()
            .assertError(FileNameMissingException::class.java)
    }

    @Test
    fun execute_emitsFileNameMissingException_whenFileNameIsEmpty() {
        useCase(NAME_EMPTY, BITMAP)
            .test()
            .assertError(FileNameMissingException::class.java)
    }

    @Test
    fun execute_successfullyCompletes_whenValidArgument() {
        whenever(fileRepository.putImageToDownloads(any(), any())) doReturn Completable.complete()
        useCase(NAME, BITMAP)
            .test()
            .assertComplete()
    }

    companion object {
        private const val NAME = "NAME"
        private const val NAME_EMPTY = ""
        private val BITMAP: Bitmap = mock()
    }
}
